CombinatorialGames/Game/Computability/FGame.lean

```diff
@@ -0,0 +1,418 @@
+
+alias ⟨_, mk_eq⟩ := mk_eq_mk
+alias _root_.SGame.Identical.mk_eq := mk_eq
+
+@[cases_eliminator]
+theorem ind {motive : FGame → Prop} (H : ∀ y, motive (mk y)) (x : FGame) : motive x :=
+  Quotient.ind H x
+
+instance : DecidableEq FGame := Quotient.decidableEq (d := SGame.Identical.instDecidable)
+
+/-- Choose an element of the equivalence class using the axiom of choice. -/
+noncomputable def out (x : FGame) : SGame := Quotient.out x
+@[simp] theorem out_eq (x : FGame) : mk x.out = x := Quotient.out_eq x
+
+/-- The finset of left moves of the game. -/
+def leftMoves : FGame → Finset FGame :=
+  Quotient.lift (fun x ↦ Finset.univ.image fun y ↦ mk (x.moveLeft y)) fun x y h ↦ by
+    ext z
+    simp_rw [Finset.mem_image, Finset.mem_univ, true_and]
+    constructor <;> rintro ⟨i, rfl⟩
+    · obtain ⟨j, hj⟩ := h.moveLeft i
+      exact ⟨j, hj.mk_eq.symm⟩
+    · obtain ⟨j, hj⟩ := h.moveLeft_symm i
+      exact ⟨j, hj.mk_eq⟩
+
+/-- The finset of right moves of the game. -/
+def rightMoves : FGame → Finset FGame :=
+  Quotient.lift (fun x ↦ Finset.univ.image fun y ↦ mk (x.moveRight y)) fun x y h ↦ by
+    ext z
+    simp_rw [Finset.mem_image, Finset.mem_univ, true_and]
+    constructor <;> rintro ⟨i, rfl⟩
+    · obtain ⟨j, hj⟩ := h.moveRight i
+      exact ⟨j, hj.mk_eq.symm⟩
+    · obtain ⟨j, hj⟩ := h.moveRight_symm i
+      exact ⟨j, hj.mk_eq⟩
+
+@[simp] theorem leftMoves_mk (x : SGame) :
+    leftMoves (mk x) = Finset.univ.image (mk ∘ x.moveLeft) := rfl
+@[simp] theorem rightMoves_mk (x : SGame) :
+    rightMoves (mk x) = Finset.univ.image (mk ∘ x.moveRight) := rfl
+
+@[ext]
+theorem ext {x y : FGame} (hl : x.leftMoves = y.leftMoves) (hr : x.rightMoves = y.rightMoves) :
+    x = y := by
+  cases x with | H x =>
+  cases y with | H y =>
+  dsimp at hl hr
+  refine (SGame.identical_iff.2 ⟨⟨?_, ?_⟩, ⟨?_, ?_⟩⟩).mk_eq <;> intro i
+  · have := Finset.mem_image_of_mem mk (show x.moveLeft i ∈ Finset.univ.image x.moveLeft by simp)
+    rw [Finset.image_image, hl] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb.symm⟩
+  · have := Finset.mem_image_of_mem mk (show y.moveLeft i ∈ Finset.univ.image y.moveLeft by simp)
+    rw [Finset.image_image, ← hl] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb⟩
+  · have := Finset.mem_image_of_mem mk (show x.moveRight i ∈ Finset.univ.image x.moveRight by simp)
+    rw [Finset.image_image, hr] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb.symm⟩
+  · have := Finset.mem_image_of_mem mk (show y.moveRight i ∈ Finset.univ.image y.moveRight by simp)
+    rw [Finset.image_image, ← hr] at this
+    obtain ⟨b, -, hb⟩ := Finset.mem_image.mp this
+    exact ⟨b, mk_eq_mk.mp hb⟩
+
+/-- `IsOption x y` means that `x` is either a left or a right move for `y`. -/
+@[aesop simp]
+def IsOption (x y : FGame) : Prop :=
+  x ∈ y.leftMoves ∨ x ∈ y.rightMoves
```

Comment:
Small golf (these predicates are def-eq)
```suggestion
  x ∈ y.leftMoves ∪ y.rightMoves
```

Reply:
They're def-eq for Set, but not for Finset - `_ ∈ _ ∨ _` is the normal form for Finset (see [`Finset#mem_union`](https://github.com/leanprover-community/mathlib4/blob/98e42c440af63586ea077289dbe8eadd1dbc5fe2/Mathlib/Data/Finset/Lattice/Basic.lean#L97-L99)) (and [`Multiset#mem_ndunion`](https://github.com/leanprover-community/mathlib4/blob/98e42c440af63586ea077289dbe8eadd1dbc5fe2/Mathlib/Data/Multiset/FinsetOps.lean#L141-L143))